src/Voron/Impl/SliceSmallSet.cs

```diff
@@ -46,10 +46,16 @@ public SliceSmallSet(int size = 0)
 
         private IEnumerable<TValue> ReturnValues()
         { 
-            if (_overflowStorage == null)
+            if (_currentIdx < _length)
             {
-                for (int i = 0; i <= _currentIdx; i++)
-                    yield return _values[i];
+                for (int i = 0; i < _length; i++)
+                {
+                    // RavenDB-20947: This may be the case of the "Cannot add a value in a read only transaction on $Root in Read"
+                    // If we dont check for 'HasValue' or that the key size is bigger than zero, we may be returning a removed
+                    // or a tree belonging to a different transaction if the array has not been properly cleaned. 
```

Comment:
Yes, I can see such a potential path to happen. In fact, that's why I changed the behavior of the LRU cached array because dealing with that path would have make it much complex. Under reset the storage would still exit which would have caused mixing old with new data.